SSA/Projects/InstCombine/ForMathlib.lean

```diff
@@ -424,6 +425,45 @@ theorem shl_ushr_eq_and_shl {w : Nat} {x C : Bitvec w.succ} :
   { rw [List.get?_eq_get] <;> simp [Nat.lt_succ_iff, *] }
   { rw [List.get?_eq_none.2] <;> simp [Nat.succ_le_iff, not_le, *] at *; assumption }
 
+-- A lot of this should probably go to a differet file here and not Mathlib
+inductive Refinement {w : Nat} : Option (Bitvec w) → Option (Bitvec w) → Prop
+  | bothSome {x y : Bitvec w } : x = y → Refinement (some x) (some y)
+  | noneAny {x? : Option (Bitvec w)} : Refinement none x?
+
+theorem Refinement.refl {w : Nat} :∀ x : Option (Bitvec w), Refinement x x := by 
+  intro x
+  cases x
+  apply Refinement.noneAny
+  apply Refinement.bothSome; rfl
+
+theorem Refinement.trans {w : Nat} : ∀ x y z : Option (Bitvec w), Refinement x y → Refinement y z → Refinement x z := by
+  intro x y z h₁ h₂
+  cases h₁ <;> cases h₂ <;> try { apply Refinement.noneAny } ; try {apply Refinement.bothSome; assumption}
+  rename_i x y hxy y h 
+  rw [hxy, h]; apply Refinement.refl
+
+instance {w : Nat} : DecidableEq (Bitvec w) := inferInstance
+instance {w : Nat} : DecidableRel (@Refinement w) := by
+  intro x y
+  cases x <;> cases y
+  { apply isTrue; exact Refinement.noneAny}
+  { apply isTrue; exact Refinement.noneAny }
+  { apply isFalse; intro h; cases h }
+  { rename_i val val'
+    cases (decEq val val')
+    { apply isFalse; intro h; cases h; contradiction } 
+    { apply isTrue; apply Refinement.bothSome; assumption }
+  }
+  
+
+instance {w : Nat} : LE (Option (Bitvec w)) := ⟨Refinement⟩
```

Comment:
My mathlib brain says not to make these global instances because there are maybe other orders it could be and it's not clear what `LE` is. I would just define a different notation for it and generalize `refinement` to `Option A`